### ISSUE TITLE
How to do multi-label classification?

### ISSUE THREAD

**Issue by KrystalCWT** [ultralytics/yolov5#1435]
Hi ,
Do you have any suggestion to modify the single-label classfication to multi-label?

**Comment by glenn-jocher** [comment]
@KRYSTALLJY multi-label classification is enabled by default in YOLOv5.

**Comment by ZeKunZhang1998** [comment]
> @KRYSTALLJY multi-label classification is enabled by default in YOLOv5.

Hello,where is the example?

**Comment by glenn-jocher** [comment]
@ZeKunZhang1998 it's automatically built in, there's no demo. You can see this here. If the model supports more than 1 class then multi-label classification is enabled. No action is required on your part.
https://github.com/ultralytics/yolov5/blob/df0e408962f010053bf5ce61e3045dcbadcd09f4/utils/general.py#L280

**Comment by KrystalCWT** [comment]
> @ZeKunZhang1998 it's automatically built in, there's no demo. You can see this here. If the model supports more than 1 class then multi-label classification is enabled. No action is required on your part.
> https://github.com/ultralytics/yolov5/blob/df0e408962f010053bf5ce61e3045dcbadcd09f4/utils/general.py#L280

but in lne 330 
i = torch.ops.torchvision.nms(boxes, scores, iou_thres)
it forces each box to have only one label. becasue nms reduce the boxes within a certain iou threshold.

**Comment by glenn-jocher** [comment]
multiple boxes are used, one per class.

**Comment by segalinc** [comment]
How would the label.txt file be like for multi-label?

**Comment by sinano1107** [comment]
@glenn-jocher 
Do you have any suggestions for assigning multiple labels to a single box?

**Comment by glenn-jocher** [comment]
@sinano1107 multi-label is enabled by default in train.py and test.py. You can enable it during inference by passing `multi_label=True` to the NMS function:

https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/utils/general.py#L417

**Comment by kazimpal87** [comment]
@glenn-jocher sorry to resurrect this. I've been looking a bit at the multi-label training and have noticed that classes which occur together tend to have much lower prediction scores, so I think there is a problem with the loss function, specifically the classification loss.

The way its formulated now is
https://github.com/ultralytics/yolov5/blob/ae4261c7749ff644f45c66b79ecb1fff06437052/utils/loss.py#L139-L142

Lets say (for the sake of example) your data has 5 classes, and there is one object in the image which belongs to both class 0 and class 3, then t will contain:
```
[[1,0,0,0,0],
 [0,0,0,1,0]]
```
and `ps[:, 5:]` will contain some logits, which will be the same in each row. When the BCEWithLogits loss is computed on this it will incorrectly penalise the logits at `[0,3]` and`[1,0]`.

If instead t was
```
[[1,0,0,1,0]]
```
then this wouldn't happen.

I'm trying to figure out a good way to solve this, but am not really sure right now.

**Comment by glenn-jocher** [comment]
@kazimpal87 yes I see your point. This is a good catch. I'm not sure of an easy solution either at the moment.

**Comment by utkutpcgl** [comment]
Maybe we can merge detections in the same bbox as a postprocessing step (if we have separate lines per class). 
Having multiple labels per line would be better but. Such as:
`(1,5,6) 0.303125 0.44583333333333336 0.10416666666666667 0.07833333333333334`

I had the same issue, people should be careful for now. **Even though the inference is multi-label, training is not**. One-hot labels reduce class confidence. It works still, but not as well as n_hot labels.

**Comment by WangRongsheng** [comment]
> @glenn-jocher sorry to resurrect this. I've been looking a bit at the multi-label training and have noticed that classes which occur together tend to have much lower prediction scores, so I think there is a problem with the loss function, specifically the classification loss.
> 
> The way its formulated now is
> 
> https://github.com/ultralytics/yolov5/blob/ae4261c7749ff644f45c66b79ecb1fff06437052/utils/loss.py#L139-L142
> 
> Lets say (for the sake of example) your data has 5 classes, and there is one object in the image which belongs to both class 0 and class 3, then t will contain:
> 
> ```
> [[1,0,0,0,0],
>  [0,0,0,1,0]]
> ```
> 
> and `ps[:, 5:]` will contain some logits, which will be the same in each row. When the BCEWithLogits loss is computed on this it will incorrectly penalise the logits at `[0,3]` and`[1,0]`.
> 
> If instead t was
> 
> ```
> [[1,0,0,1,0]]
> ```
> 
> then this wouldn't happen.
> 
> I'm trying to figure out a good way to solve this, but am not really sure right now.

请问这么做能否有效进行多标签分类模型的训练和推理？

**Comment by lijianyu1985** [comment]
> > @glenn-jocher sorry to resurrect this. I've been looking a bit at the multi-label training and have noticed that classes which occur together tend to have much lower prediction scores, so I think there is a problem with the loss function, specifically the classification loss.
> > The way its formulated now is
> > https://github.com/ultralytics/yolov5/blob/ae4261c7749ff644f45c66b79ecb1fff06437052/utils/loss.py#L139-L142
> > 
> > Lets say (for the sake of example) your data has 5 classes, and there is one object in the image which belongs to both class 0 and class 3, then t will contain:
> > ```
> > [[1,0,0,0,0],
> >  [0,0,0,1,0]]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and `ps[:, 5:]` will contain some logits, which will be the same in each row. When the BCEWithLogits loss is computed on this it will incorrectly penalise the logits at `[0,3]` and`[1,0]`.
> > If instead t was
> > ```
> > [[1,0,0,1,0]]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then this wouldn't happen.
> > I'm trying to figure out a good way to solve this, but am not really sure right now.
> 
> 请问这么做能否有效进行多标签分类模型的训练和推理？

同问

**Comment by thiagodma** [comment]
> @glenn-jocher sorry to resurrect this. I've been looking a bit at the multi-label training and have noticed that classes which occur together tend to have much lower prediction scores, so I think there is a problem with the loss function, specifically the classification loss.
> 
> The way its formulated now is
> 
> https://github.com/ultralytics/yolov5/blob/ae4261c7749ff644f45c66b79ecb1fff06437052/utils/loss.py#L139-L142
> 
> Lets say (for the sake of example) your data has 5 classes, and there is one object in the image which belongs to both class 0 and class 3, then t will contain:
> 
> ```
> [[1,0,0,0,0],
>  [0,0,0,1,0]]
> ```
> 
> and `ps[:, 5:]` will contain some logits, which will be the same in each row. When the BCEWithLogits loss is computed on this it will incorrectly penalise the logits at `[0,3]` and`[1,0]`.
> 
> If instead t was
> 
> ```
> [[1,0,0,1,0]]
> ```
> 
> then this wouldn't happen.
> 
> I'm trying to figure out a good way to solve this, but am not really sure right now.

Hi @kazimpal87, have you found a solution to this?

**Comment by glenn-jocher** [comment]
@thiagodma Hi, have you found a solution to this issue?

**Comment by thiagodma** [comment]
Hey @glenn-jocher ! No, I haven't... As a workaround, I'm using lower thresholds

**Comment by glenn-jocher** [comment]
Thank you for sharing your workaround, @thiagodma! Lowering the thresholds can indeed help mitigate the issue. If you have any further insights or developments, feel free to share them here.
